File: mobile/src/memory_log.py
"""
memory_log.py
Mobile equivalent of web/src/memoryLog.js

Responsibility:
  - store_step(log_list, step_data) — append a validated step entry
  - save_log(log_list, file_path)   — write log to JSON file
  - load_log(file_path)             — read existing log from JSON file

Schema per entry (IDENTICAL to web memory_log.json schema):
{
  "step":                   int,
  "from_url":               str,   # mobile: "package/activity"
  "from_title":             str,   # mobile: package name
  "action":                 str,   # "tap" | "type" | "swipe" | "error"
  "target":                 str,   # resource_id or class name
  "target_element_details": {
    "elementId":   int | None,
    "tag":         str,
    "text":        str,
    "id":          str | None,     # resource_id mapped to "id" for schema compat
    "class":       str,
    "selector":    str             # resource_id used as selector
  } | None,
  "to_url":             str,
  "to_title":           str,
  "screenshot_before":  str,       # relative path
  "screenshot_after":   str,       # relative path
  "timestamp":          str        # ISO 8601
}
"""

import json
import os
from datetime import datetime, timezone


def store_step(log_list: list, step_data: dict) -> None:
    """
    Appends a validated step entry to the in-memory log list.
    Mirrors storeStep() in memoryLog.js — same field names for schema compatibility.

    Args:
        log_list:  The shared in-memory log list
        step_data: Raw step data dict
    """
    if not isinstance(log_list, list):
        raise TypeError("log_list must be a list")

    ted = step_data.get("target_element_details")
    if ted:
        # Map mobile fields → web-compatible schema fields
        normalised_ted = {
            "elementId": ted.get("elementId"),
            "tag":       ted.get("tag", ""),
            "text":      ted.get("text", ""),
            "id":        ted.get("resource_id") or ted.get("id"),   # resource_id → id
            "class":     ted.get("class_name") or ted.get("class", ""),
            "selector":  ted.get("resource_id") or ted.get("selector", ""),
        }
    else:
        normalised_ted = None

    entry = {
        "step":                   step_data.get("step", len(log_list)),
        "from_url":               step_data.get("from_url", ""),
        "from_title":             step_data.get("from_title", ""),
        "action":                 step_data.get("action", "unknown"),
        "target":                 step_data.get("target", ""),
        "target_element_details": normalised_ted,
        "to_url":                 step_data.get("to_url", ""),
        "to_title":               step_data.get("to_title", ""),
        "screenshot_before":      step_data.get("screenshot_before", ""),
        "screenshot_after":       step_data.get("screenshot_after", ""),
        "timestamp":              step_data.get("timestamp", datetime.now(timezone.utc).isoformat()),
    }

    # Optional fields — only include if present (mirrors web side spread operator)
    if step_data.get("error"):
        entry["error"] = step_data["error"]
    if step_data.get("value"):
        entry["value"] = step_data["value"]

    log_list.append(entry)


def save_log(log_list: list, file_path: str) -> None:
    """
    Serialises the full log list to a JSON file.
    Creates parent directories if they don't exist.
    Mirrors saveLog() in memoryLog.js.

    Args:
        log_list:  The in-memory log list
        file_path: Destination file path (e.g. 'logs/mobile_memory_log.json')
    """
    os.makedirs(os.path.dirname(file_path), exist_ok=True)
    with open(file_path, "w", encoding="utf-8") as f:
        json.dump(log_list, f, indent=2)
    print(f"[memory_log] Saved {len(log_list)} step(s) → {file_path}")


def load_log(file_path: str) -> list:
    """
    Reads and parses an existing memory_log.json file.
    Returns empty list if file does not exist.
    Mirrors loadLog() in memoryLog.js.

    Args:
        file_path: Path to the JSON log file

    Returns:
        Parsed list of step dicts, or [] on missing/corrupt file
    """
    if not os.path.exists(file_path):
        return []
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            return json.load(f)
    except (json.JSONDecodeError, OSError) as e:
        print(f"[memory_log] Failed to load {file_path}: {e}")
        return []

File: web/src/memoryLog.js
/**
 * memoryLog.js
 * Day 1 – Nidhi K
 * Responsibility: In-memory log management for exploration steps.
 * storeStep() appends a step entry; saveLog() writes the full log to JSON.
 *
 * Schema per entry:
 * {
 *   step: number,
 *   from_url: string,
 *   from_title: string,
 *   action: 'click' | 'fill' | 'navigate',
 *   target: string,           // element tag name or URL
 *   target_element_details: {
 *     elementId: number,
 *     tag: string,
 *     text: string,
 *     id: string | null,
 *     class: string,
 *     selector: string
 *   },
 *   to_url: string,
 *   to_title: string,
 *   screenshot_before: string,   // relative path
 *   screenshot_after: string,    // relative path
 *   timestamp: string            // ISO 8601
 * }
 */

const fs = require('fs');
const path = require('path');

/**
 * storeStep(logArray, stepData)
 * Pushes a validated step entry into the in-memory log array.
 *
 * @param {Array} logArray        - The shared in-memory log array
 * @param {Object} stepData       - Raw step data object (see schema above)
 * @returns {void}
 */
function storeStep(logArray, stepData) {
  if (!Array.isArray(logArray)) {
    throw new TypeError('logArray must be an array');
  }

  // Normalise + validate required fields with safe defaults.
  // Legacy keys (step/from_url/action/target/...) are preserved for backward
  // compatibility; new structured keys support the Fusion observation schema.
  const entry = {
    step: typeof stepData.step === 'number' ? stepData.step : logArray.length,
    state_id: stepData.state_id || null,
    parent_state_id: stepData.parent_state_id || null,
    from_url: stepData.from_url || '',
    from_title: stepData.from_title || '',
    action: stepData.action || 'unknown',
    action_details: stepData.action_details || {
      type: stepData.action || 'unknown',
      target: typeof stepData.target === 'string' ? stepData.target : '',
      value: stepData.value || '',
    },
    target: stepData.target || '',
    target_element_details: stepData.target_element_details
      ? {
          elementId: stepData.target_element_details.elementId ?? null,
          tag: stepData.target_element_details.tag || '',
          text: stepData.target_element_details.text || '',
          id: stepData.target_element_details.id || null,
          class: stepData.target_element_details.class || stepData.target_element_details.className || '',
          selector: stepData.target_element_details.selector || '',
        }
      : null,
    result: {
      url_after: stepData.to_url || '',
      success: stepData.success !== undefined ? Boolean(stepData.success) : true,
      error: stepData.error || null,
    },
    to_url: stepData.to_url || '',
    to_title: stepData.to_title || '',
    elements_observed: stepData.elements_observed ?? null,
    screenshot_before: stepData.screenshot_before || '',
    screenshot_after: stepData.screenshot_after || '',
    timestamp: stepData.timestamp || new Date().toISOString(),
  };

  logArray.push(entry);
}

/**
 * saveLog(logArray, filePath)
 * Serialises the full log array to a JSON file.
 * Creates parent directories if they don't exist.
 *
 * @param {Array}  logArray  - The in-memory log array
 * @param {string} filePath  - Destination file path (e.g. 'logs/memory_log.json')
 * @returns {void}
 */
function saveLog(logArray, filePath) {
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
  fs.writeFileSync(filePath, JSON.stringify(logArray, null, 2), 'utf8');
  console.log(`[memoryLog] Saved ${logArray.length} step(s) → ${filePath}`);
}

/**
 * loadLog(filePath)
 * Reads and parses an existing memory_log.json file.
 * Returns empty array if file does not exist.
 *
 * @param {string} filePath
 * @returns {Array}
 */
function loadLog(filePath) {
  if (!fs.existsSync(filePath)) return [];
  try {
    const raw = fs.readFileSync(filePath, 'utf8');
    return JSON.parse(raw);
  } catch (err) {
    console.error(`[memoryLog] Failed to load ${filePath}:`, err.message);
    return [];
  }
}

/**
 * saveStates(states, filePath) — persist the exploration state records.
 * Each state: {state_id, parent_state_id, url, title, fingerprint,
 *              elements_observed, leading_action, timestamp}
 */
function saveStates(states, filePath) {
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(filePath, JSON.stringify(states, null, 2), 'utf8');
  console.log(`[memoryLog] Saved ${states.length} state(s) → ${filePath}`);
}

/**
 * saveTransitions(transitions, filePath) — persist the edge list:
 * {from_state, action:{type,target,value}, to_state, result, url_after, timestamp}
 */
function saveTransitions(transitions, filePath) {
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(filePath, JSON.stringify(transitions, null, 2), 'utf8');
  console.log(`[memoryLog] Saved ${transitions.length} transition(s) → ${filePath}`);
}

/**
 * saveSummary(summary, filePath) — run-level machine-readable summary
 * (totals, termination reason, limits, unique URLs).
 */
function saveSummary(summary, filePath) {
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  fs.writeFileSync(filePath, JSON.stringify(summary, null, 2), 'utf8');
  console.log(`[memoryLog] Saved exploration summary → ${filePath}`);
}

module.exports = { storeStep, saveLog, loadLog, saveStates, saveTransitions, saveSummary };
